src/Microsoft.ML.FastTree/TreeEnsemble/InternalRegressionTree.cs

```diff
@@ -1515,28 +1515,59 @@ public void AppendFeatureContributions(in VBuffer<float> src, BufferBuilder<floa
             int node = 0;
             while (node >= 0)
             {
-                int ifeat = SplitFeatures[node];
-                var val = src.GetItemOrDefault(ifeat);
-                val = GetFeatureValue(val, node);
                 int otherWay;
-                if (val <= RawThresholds[node])
+                if (CategoricalSplit[node])
                 {
+                    Contracts.Assert(CategoricalSplitFeatures != null);
+
+                    int newNode = LteChild[node];
                     otherWay = GtChild[node];
-                    node = LteChild[node];
+                    foreach (var index in CategoricalSplitFeatures[node])
+                    {
+                        float fv = GetFeatureValue(src.GetItemOrDefault(index), node);
+                        if (fv > 0.0f)
+                        {
+                            newNode = GtChild[node];
+                            otherWay = LteChild[node];
+                            break;
+                        }
+                    }
+
+                    // What if we went the other way?
+                    var ghostLeaf = GetLeafFrom(in src, otherWay);
+                    var ghostOutput = GetOutput(ghostLeaf);
+
+                    // If the ghost got a smaller output, the contribution of the categorical features is positive, so
+                    // the contribution is true minus ghost.
+                    foreach(var ifeat in CategoricalSplitFeatures[node])
+                        contributions.AddFeature(ifeat, (float)(trueOutput - ghostOutput));
```

Comment:
Code-wise, I guess this is the correct way to find which features are involved in the categorical split of this node (as it is done in a similar way [here](https://github.com/dotnet/machinelearning/blob/1ea53362a692e551737341787d7b49541c28dfb8/src/Microsoft.ML.FastTree/TreeEnsemble/InternalRegressionTree.cs#L851-L877))

But I don't know if this is the "mathematically correct" way of calculating feature contributions for categorical features.

Reply:
So I've updated this. I am still not sure the updated version is correct. But I think it's the closest to how feature contribution is calculated for numerical feature splits, and considering that for other other cases (FastTree, Gam, etc...) categorical features are treated the same as any other feature (ignoring the fact they're categorical) when calculating feature contribution.